[APM] Moving the date picker into APM

jasonrhodes opened the pull request:
Closes elastic/kibana#30388 
Closes elastic/kibana#32025

This PR removes the angular date picker entirely and pulls in the `EuiSuperDatePicker` as a flex element to the right of the existing "kuery bar".

Here is how it looks without the refresh/update button, as mentioned by @formgeist [in the comments below](https://github.com/elastic/kibana/pull/31311#issuecomment-465098920). [As noted there](https://github.com/elastic/kibana/pull/31311#issuecomment-465100861), this results in more network requests as users interact with the absolute dates/times.

<img width="1438" alt="screen shot 2019-02-19 at 6 55 32 am" src="https://user-images.githubusercontent.com/159370/53013317-7dbbe680-3413-11e9-86b9-73be16d1cacf.png">

jasonrhodes committed:
WIP moving the date picker into APM

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/967/)

sqren commented on x-pack/plugins/apm/public/store/urlParams.ts:
```diff
@@ -138,8 +138,13 @@ function getPathParams(pathname: string) {
-export function updateTimePicker(time: string) {
+export function updateTimePicker(time: TimeUpdate) {
   return { type: TIMEPICKER_UPDATE, time };
 }
```
We can get rid of redux entirely when updating the time. `onTimeChange` should simply change the url.

jasonrhodes replied:
Hmm, I think we still need to dispatch with the datemath-parsed values so they can be assigned to `urlParams.start` and `urlParams.end` 

https://github.com/elastic/kibana/blob/a0bc7dae55694d99873e511fb8e0ac0bbd419338/x-pack/plugins/apm/public/store/urlParams.ts#L80

Or else we need to do the datemath parsing inside the location update reducer, I think?

jasonrhodes replied:
Oh also: during auto-refresh, we often don't need to update the URL (because the URL is a relative date string) but we will need to recalculate the parsed datemath values and update them in the store, so I think the TIMEPICKER_UPDATE is still necessary.

sqren replied:
> Oh also: during auto-refresh, we often don't need to update the URL (because the URL is a relative date string) but we will need to recalculate the parsed datemath values and update them in the store, so I think the TIMEPICKER_UPDATE is still necessary.

Good point - didn't think of that.

sqren commented on x-pack/plugins/apm/public/utils/timepicker/index.js:
```diff
@@ -24,6 +24,7 @@ const waitForAngularReady = new Promise(resolve => {
+  return Promise.resolve();
   return new Promise(resolve => {
     // default the timepicker to the last 24 hours
     chrome.getUiSettingsClient().overrideLocalDefault(
```
This entire file can be deleted now, right?

jasonrhodes replied:
Yeah, once we're sure it all works.

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,220 @@
+      ...this.props.location,
+      search: fromQuery(nextSearch)
+    });
+  }
```
I think we should very seriously consider getting rid of rison inside APM UI. It will simplify url parsing a lot. We still need rison when handling urls to Discover but we can treat that as legacy.

jasonrhodes replied:
Are we sure there aren't any incoming links that we care about breaking if we don't read from the rison on load, at least? Does Kibana really want every app handling date/time in their own way with different query params in every app?

sqren replied:
> Are we sure there aren't any incoming links that we care about breaking if we don't read from the rison on load, at least?

I can't think of other plugins than Infra that links to us. We should mention this to them.

> Does Kibana really want every app handling date/time in their own way with different query params in every app?

In an ideal world, no. But I think we should use this opportunity to simplify something that has bothered us for a while. And then we can hopefully affect the rest of the platform over time.

jasonrhodes replied:
@makwarth do you have any opinion on this? Options:

1. Leave the URL format as it works today, which means reading from/writing to the URL in this "rison" format which involves some unnecessary complexity around encoding and decoding the time range and refresh values. Any inbound links into APM that specify date ranges would continue to work the same way the do today.

1. Begin using our own query parameters outside of the existing `_g` param, without the "rison" format. Simplifies our code reading from and writing to the URL. Downside is we could break inbound links, and it increases the complexity of the work being done in this ticket a bit. We also need to pick a new set of query parameters which would be nice to coordinate with other Kibana apps, but there may not be time to do that coordination at this point in the release cycle. 

Thoughts? I think we should definitely get rid of the rison complexity, but I wasn't sure if that was something we wanted to do right now in this ticket.

formgeist replied:
I imagine we'd want to go with option 1 for now, for the risks pointed out in option 2, but also because of two reasons;
 
1. We only want to use the UI component and not bring with it additional refactoring of the time handling if possible. 

2. I assume there'll be changes incoming from the platform team when/if they're addressing the date and time handling. Also assuming we're not the only team dependent on the rison format?

sqren replied:
> 1. We only want to use the UI component and not bring with it additional refactoring of the time handling if possible.

My point is that Rison has a lot of complexity, which is visible in this PR. I'm convinced that this PR could be smaller, if we dropped the Rison dep.

> 2\. I assume there'll be changes incoming from the platform team when/if they're addressing the date and time handling. Also assuming we're not the only team dependent on the rison format?

Just to clear up a potential misunderstanding: The Timepicker UI Component or the platform doesn't prescribe any specific date format. We can choose to use Rison encoded values for time range, as well as for all the other values, or not. Eg. ML went full-on Rison for all values. We have only used Rison for date ranges, and normal encoding for other values.

sqren replied:
@jasonrhodes I was only able to find this inbound link to APM, and it doesn't contain time range:

https://github.com/elastic/kibana/blob/00fbc2572efe519dd2169b3e21845b1397ecdf21/x-pack/plugins/infra/public/components/waffle/node_context_menu.tsx#L54-L63

formgeist replied:
@sqren My response was merely to says that I'm in favour of leaving things a little more complex until we can perhaps refactor things to depend more or less on Rison for another release. I'm not sure I understand why there's no suggested way to handle date formatting in Kibana, but I'm also sure someone once gave me the answer to this and I've forgotten it (when we first implemented APM UI).

sqren replied:
> until we can perhaps refactor things to depend more or less on Rison for another release

That would be in 8.0 to avoid breaking the urls. I don't think I can handle Rison in APM UI that long 😅

jasonrhodes replied:
Rison has been removed -- let me know if anyone objects. :) Resolving this for now.

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,220 @@
+    to: '',
+    isPaused: true,
+    refreshInterval: 0
+  };
```
Why is this component stateful? Shouldn't it just read the props from urlParams (where we can also set the defaults)?

jasonrhodes replied:
That's possibly true, the state was just set up to get the EUI component to display correctly since it's controlled from here. But the query state may be enough to keep it synced, I'll take a look.

jasonrhodes replied:
Ah right, while we're still using rison, we need to store the decoded values in state or else decode them in every render.

sqren replied:
> Ah right, while we're still using rison, we need to store the decoded values in state or else decode them in every render.

Did you find any perf issues? Else it sounds a little like premature optimisation. 
I hope we can get rid of Rison, and even if we can't I think we should keep the parsing of url params consistent with all the other url params.

jasonrhodes replied:
Hm ok. Not sure about premature optimization, it just seems a bit weird to do the decode in every render and also in every refresh cycle. I’ll wait until we make a final decision on rison to worry about this though. 

> I think we should still keep the parsing of url params consistent with all the other url params - and then we can apply memoization where needed.

Do you mean by storing them in the redux store? We don't currently store any of this local datepicker-specific info in redux -- `urlParams.start` and `urlParams.end` are the calculated date values used by all the ES queries, but these date values need to be the raw strings found in `from` and `to` so they can be passed into the date picker component that way. We don't use those raw strings anywhere else in the app, and we don't use the pause and refresh interval state anywhere else either, fwiw. 

Once we sort the rison question I'll revisit this and see if we can remove the local state.

sqren replied:
> We don't currently store any of this local datepicker-specific info in redux

No, but don't you think that's a odd when we handle the other url params in the reducer?

sqren replied:
> Once we sort the rison question I'll revisit this and see if we can remove the local state

What do you think about trying it out, and seeing if it simplifies things, and then make a decision?

jasonrhodes replied:
> don't you think that's odd when we handle the other url params in the reducer?

I get what you're saying here re: consistency, but it also seems odd to move things into Redux that are only ever used in this one local component, and a bit complicated to manage `urlParams.start` next to `urlParams.from` etc. Not sure which odd thing is better here… let me think on it and we can chat in stand up.

jasonrhodes committed:
Stable version of EUI date picker controls, still needs layout

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1247/)

jasonrhodes committed:
Flex layout for kuery bar and date picker

jasonrhodes committed:
Removes angular time picker logic and layout

formgeist commented:
Perhaps wrong place to drop this in, but as we addressed filtering (#30943), I'm thinking of setting `showUpdateButton` to `false` since we don't need it to take up prime real estate in that horizontal area. Would it be possible to include in this PR?

jasonrhodes commented:
> Perhaps wrong place to drop this in, but as we addressed filtering (#30943), I'm thinking of setting `showUpdateButton` to `false` since we don't need it to take up prime real estate in that horizontal area. Would it be possible to include in this PR?

Yep, this is easy. To be clear, removing that button results in a lot more network requests as people interact with the dates and times, but if we're okay with that then this is just a single prop change on the picker.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1278/)

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,210 @@
+      this.props.dispatchUpdateTimePicker({ min, max });
+      this.lastRefresh = now;
+    }
+    requestAnimationFrame(() => this.refresh(min, max));
```
Why is `raf` needed?

jasonrhodes replied:
@ogupte and I talked about the dangers of using `setInterval` last week, and we were flipping between this and a recursive `setTimeout`. Still thinking about which way I want to go there.

jasonrhodes replied:
Removed raf in favor of recursive setTimeout

jasonrhodes committed:
Fixes snapshot test

jasonrhodes committed:
Adds integration tests for date picker

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1314/)

jasonrhodes commented:
@sqren I took a stab at some less-mocked style "integration" tests for this DatePicker component -> https://github.com/elastic/kibana/pull/31311/commits/e82415b1ca39e9c0d55e2d6bf2b1c274c4551c3d 

Let me know what you think.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1327/)

jasonrhodes committed:
Simplifies refresh cycle with setTimeout over requestAnimationFrame

jasonrhodes committed:
Removes rison and local state from APM date picker flow

jasonrhodes commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,219 @@
+
+  public clearRefreshTimeout() {
+    if (this.refreshTimeoutId) {
+      window.clearTimeout(this.refreshTimeoutId);
```
Explicitly using `window` here and for `window.setTimeout` below because Typescript otherwise assumes you are using the NodeJS implementation which returns a special "NodeJS.Timeout" typed value rather than just a number. 

We could alternatively use "as number" in a few places to tell TS what we want, but this seemed ok since Jest runs tests with JSDom.

sqren replied:
👍

jasonrhodes commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,219 @@
+    return {
+      ...APM_DEFAULT_TIME_OPTIONS,
+      ...converted
+    } as DatePickerParams;
```
TS doesn't understand that by defaulting to an object with all of the values defined, none of the four values in this interface can be undefined, so I've used `as` here to help TS understand.

sqren replied:
I don't think it's TS that can't handle this, rather that I have modelled it incorrectly in url_helpers. 
It can probably be done smarter, but this is at least a start:

```ts
interface APMQueryParamsRaw {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string;
  refreshInterval?: string;
}

interface APMQueryParamsBoth {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string | number;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string | boolean;
  refreshInterval?: string | number;
}
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1388/)

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,215 @@
+}
+
+// TODO this can be removed when EUI PR lands: https://github.com/elastic/eui/pull/1574
+declare module '@elastic/eui' {
```
Until then, should we have it in the "typings" folder?

sqren replied:
Eg. like this: https://github.com/elastic/kibana/blob/00fbc2572efe519dd2169b3e21845b1397ecdf21/x-pack/plugins/apm/typings/numeral.d.ts

sqren commented on x-pack/plugins/apm/public/components/shared/Links/url_helpers.ts:
```diff
@@ -127,6 +127,10 @@ interface APMQueryParams {
   kuery?: string;
+  rangeFrom?: string;
+  rangeTo?: string;
+  refreshPaused?: string | boolean;
```
A query param is always a string afaik. Any booleans would be stringified like `"true"`

jasonrhodes replied:
Right, but if I leave this as string only, then I have to manually convert `true` to `"true"` before I use `fromQuery({ refreshPaused: true })`. Since this interface is used for the converted query object, sometimes those values are strings when they come back from `qs.parse` and sometimes they're boolean when they are going in. 

Do you have thoughts on a better way to handle this?

sqren replied:
Yes, I suggest we fix `fromQuery` and `toQuery` in url_helpers. They are incorrect, and I never noticed until now. Something like this:


```ts
// toQuery always return an object of strings
export function toQuery(search?: string): APMQueryParamsRaw & RisonEncoded {
  return search ? qs.parse(search.slice(1)) : {};
}

// fromQuery is more liberal in what it accepts since it both accepts: `refreshPaused: true` and `refreshPaused: 'true'`
export function fromQuery(query: APMQueryParams & RisonEncoded) {
  const encodedQuery = encodeQuery(query, ['_g', '_a']);
  return stringifyWithoutEncoding(encodedQuery);
}

interface APMQueryParamsRaw {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string;
  refreshInterval?: string;
}

export interface APMQueryParams {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string | number;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string | boolean;
  refreshInterval?: string | number;
}
```

sqren replied:
It's a bit of a mess right now with all the Rison. I hope we can move that out, and only handle that in the Discover Links.

sqren commented on x-pack/plugins/apm/public/store/urlParams.ts:
```diff
@@ -138,8 +138,13 @@ function getPathParams(pathname: string) {
+
 // ACTION CREATORS
-export function updateTimePicker(time: string) {
+export function updateTimePicker(time: TimeUpdate) {
```
This should be moved to `DatePicker.tsx` if that's the only place where it's consumed

jasonrhodes replied:
I think it makes sense to leave the action creator here where the reducer is, they share the action type constant value and I don't think we have many action creators in components anywhere else. It was only being used in one place before, too.

sqren replied:
Oh yes, we still use `TIMEPICKER_UPDATE` in the reducer, in urlParams. Forgot about that.

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,215 @@
+    const converted = {
+      ...query,
+      refreshPaused:
+        query.refreshPaused === 'true' || query.refreshPaused === true,
```
Value will always be a string, right?
```js
refreshPaused: query.refreshPaused === 'true'
```

This is one of the reasons I do all conversion in urlParams eg:
https://github.com/jasonrhodes/kibana/blob/e6f30167d67f030c180b098d5f3a4d20fa15622b/x-pack/plugins/apm/public/store/urlParams.ts#L61

The rest of the application only uses `urlParams` (and not the raw url values) and thus can be sure they are the correct type.

jasonrhodes replied:
Yeah, I mean I'm not totally against using urlParams for all of this. I think it may cause a little confusion around start/end vs to/from, but if we think that's the simplest way to handle all these values I can probably get behind it. What do you think?

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,215 @@
+      refreshInterval:
+        typeof query.refreshInterval === 'undefined'
+          ? undefined
+          : Number(query.refreshInterval)
```
We already have a `toNumber` utility method. Can you use that?
https://github.com/jasonrhodes/kibana/blob/e6f30167d67f030c180b098d5f3a4d20fa15622b/x-pack/plugins/apm/public/store/urlParams.ts#L87-L92

jasonrhodes replied:
Sure. Why does that helper accept an array of strings and convert it to undefined?

sqren replied:
> Why does that helper accept an array of strings and convert it to undefined?

It's a very good question. I git blamed and found myself guilty. Unfortunately I can't remember why and I didn't add a test :( Looking at the usages of `toNumber` I think it's safe to remove the array handling, and simply do

```
export function toNumber(value?: string) {
  if (value !== undefined) {
    return parseInt(value, 10);
  }
}
```

sqren replied:
Hmm, it's actually something you did in https://github.com/elastic/kibana/pull/24294/files.
And `toString` was also changed to handle arrays. I think neither should handle arrays tbh.

sqren replied:
And while we are added:
```
kuery: legacyDecodeURIComponent(kuery as string | undefined),
```

can be changed back to:
```
kuery: legacyDecodeURIComponent(kuery),
```

jasonrhodes committed:
Adds refresh tests, fixes some refresh logic

jasonrhodes committed:
Moves temporary EUI types out of component

jasonrhodes committed:
Moves toBoolean helper and fixes TS error

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1596/)

jasonrhodes commented:
TIL `jest` has fake timers ... that might simplify some of the tests here, gonna take a stab.

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,189 @@
+      ...APM_DEFAULT_TIME_OPTIONS,
+      ...query
+    } as DatePickerParams;
+  });
```
I had a little bit of a hard time, understanding what this method was actually returning due to the conditionals and spreads. Since it's only four params I think it's worth making them explicit. What do you think about:

```ts
public getParamsFromSearch = (search: string) => {
  const { rangeFrom, rangeTo, refreshPaused, refreshInterval } = toQuery(
    search
  );
  return {
    rangeFrom: rangeFrom || 'now-24h',
    rangeTo: rangeTo || 'now',
    refreshPaused: toBoolean(refreshPaused),
    refreshInterval: toNumber(refreshInterval) || 0
  };
};
```

btw. This requires some minor modifications to `APMQueryParams` in url_helpers

jasonrhodes replied:
Hm, yeah, that function could use some work. This suggestion makes it much harder to see the defaults at a glance, and actually prevents us from being able to use a default value for `refreshPaused` which we want to default to `true`. Once I get the types sorted based on your good feedback there, I'll look at this again and see what we can do.

sqren replied:
ah, missed the default for `refreshPaused`. Maybe `toBoolean` should take a default value as a second param?

jasonrhodes replied:
OK I was actually able to move _all_ of these defaults up into the destructure, which reads nicely and avoids the confusion from before.

sqren replied:
Looks good 👍

sqren commented on x-pack/plugins/apm/public/store/urlParams.ts:
```diff
@@ -102,6 +102,15 @@ function toString(str?: string | string[]) {
+  if (value === 'false' || value === false) {
+    return false;
+  }
+}
```
I think these conversion methods should only expect the raw url param, which is always a string. In that case this is sufficient:

```js
export function toBoolean(value?: string) {
  return value === 'true';
}
```

sqren commented:
@jasonrhodes I've added a bunch of feedback regarding types. I got it to work without any type casting in a local branch, and can push that up if you'd like or we can zoom about it.

jasonrhodes commented:
> @jasonrhodes I've added a bunch of feedback regarding types. I got it to work without any type casting in a local branch, and can push that up if you'd like or we can zoom about it.

@sqren do you want to just PR against this one with your changes?

jasonrhodes committed:
Removes unused Link import

sorenlouv committed:
Types for datepicker (WIP)

elasticmachine commented:
Pinging @elastic/apm-ui

jasonrhodes committed:
Updates default date picker values after merging in Søren's type changes

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx:
```diff
@@ -0,0 +1,175 @@
+      rangeTo = 'now',
+      refreshPaused = 'true',
+      refreshInterval = '0'
+    } = toQuery(search);
```
Nice defaults 👍

jasonrhodes committed:
Streamlines new APM query types to prevent duplication

jasonrhodes committed:
Uses jest fake timers for refresh tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1828/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2009/)

sqren commented:
I think this looks good. Haven't tested it properly yet. How does the rest of the links work (Discover and ML buttons), now that Rison is gone from the url?

formgeist commented:
@jasonrhodes As per @cchaos's [comment](https://github.com/elastic/eui/issues/1599#issuecomment-467459194), we should add the refresh/apply button back in. We'll have to consider the space utilization when we're going to build the dedicated filtering at a later stage. Sorry about the confusion 👍

sorenlouv reviewed (changes requested):
Just tried this out, and as we talked about there are a few things missing:
 - Internal APM links should preserve the new non-rison time format
 - External links to Discover and ML should have applied time in Rison

jasonrhodes committed:
Updates url handling to remove Rison from APM URLs, keeps Rison in outgoing Kibana links

jasonrhodes commented:
@sqren I've addressed both link issues you mention here:

> Just tried this out, and as we talked about there are a few things missing:
> 
> * Internal APM links should preserve the new non-rison time format
> * External links to Discover and ML should have applied time in Rison

as well as adjusting the breadcrumbs so they also inherit and carry forward the new non-rison format.

I'll definitely appreciate your eyes on this in case there are gotchas I'm not thinking with how I've refactored the url_helpers file, or other corner cases that need to be addressed with URL handling in general. Thanks!

I'll work on updating and adding tests tomorrow.

jasonrhodes commented:
> @jasonrhodes As per @cchaos's [comment](https://github.com/elastic/eui/issues/1599#issuecomment-467459194), we should add the refresh/apply button back in. We'll have to consider the space utilization when we're going to build the dedicated filtering at a later stage. Sorry about the confusion 👍

Ah ok, will add this back in tomorrow as well. Thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2877/)

sqren commented:
I'm seeing some odd issue where nothing is rendered on the screen. It seems like the router doesn't match any routes.
To reproduce type the url `/app/apm#/opbeans-node/transactions` (make sure it does a full refresh, so open from a new tab or similar).
The screen is blank, and `_g=()` is appended (which it shouldn't since we are done with rison)

sqren commented on x-pack/plugins/apm/public/components/shared/Links/url_helpers.ts:
```diff
@@ -6,95 +6,89 @@
+  return qs.stringify(risonQuery, undefined, undefined, {
+    encodeURIComponent: (v: string) => v
+  });
 }
```
It would be really great if we can get Rison out of url_helpers completely, and make it a separate converter, that `DiscoverLink` and `MLJobLink` can use. This way it's separated from how urls are handled internally in APM.

jasonrhodes replied:
Mm, this is a good idea. I'll take a stab at it and ping you if I have trouble figuring out how to make this work cleanly.

sqren commented:
The problem with the blank screen is caused by `start` and `end` in urlParams being undefined.

jasonrhodes committed:
Move filter bar up and out from within a react-redux-request to avoid catch-22 circular dependency

jasonrhodes commented:
@sqren great catch on the circular depencency there. What do you think of this solution: https://github.com/elastic/kibana/pull/31311/commits/69cf79ddf43059d3420c1a028197837b91e79e56

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2922/)

sqren commented:
>  great catch on the circular dependency there. What do you think of this solution: [69cf79d](https://github.com/elastic/kibana/commit/69cf79ddf43059d3420c1a028197837b91e79e56)
@jasonrhodes 

That could definitely work. I'm still thinking having the default in urlParams is more robust, since this is where the rest of the application reads from. If, for some reason, the Timepicker doesn't render (like in this case) the app will render with the default 24h timerange.

jasonrhodes commented:
@sqren I didn't see this edited part of your note: "I think the cleanest solution would be to move the default time range from the DatePicker to urlParams.ts" until after I'd pushed my proposed solution. :/

So to do this we'll have to move the date math parsing into urlParams as well. I'll see what it looks like in a bit after I sort the rison refactor.

jasonrhodes committed:
Separates rison encoding from regular url handling

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2935/)

jasonrhodes committed:
Sets start and end defaults in urlParams reducer

jasonrhodes commented:
@sqren ok I've separated the rison stuff completely from everything else, with some duplication but I think it makes the type flows much clearer than trying to use a boolean `isRison` toggle on `KibanaLink` etc.

I've also just pushed a commit that sets default calculated start/end values in the urlParams reducer. I'm not 100% sure if this is what you were thinking when you suggested defaulting things there but this will prevent issues if the DatePicker never loads, or doesn't load early enough, etc.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2942/)

sqren commented on x-pack/plugins/apm/public/components/shared/Links/DiscoverLinks/DiscoverLink.tsx:
```diff
@@ -5,20 +5,20 @@
+        <KibanaRisonLink
           pathname={'/app/kibana'}
           hash={'/discover'}
           query={queryWithIndexPattern}
```
I think the separation is really good 👍 
This might be a bit too much for now, but I'm thinking we might be able to strip all the Rison logic even more down, so we don't need the rison_helper stuff. Pseudocode of the `DiscoverLink` component:

```js
export function DiscoverLink({
  query,
  children,
  urlParams,
  indexPattern
}: Props) {
  const risonEncoded = rison.encode({
    ...query,
    _g: {
      time: {
        from: urlParams.fromRange,
        to: urlParams.toRange
      }
    }
  });
  
  const search = risonEncoded // or something like it

  const href = url.format({
    pathname: chrome.addBasePath(pathname),
    hash: `${hash}?${search}`
  });

  return <EuiLink href={href} children={children} />;
}
```

sqren replied:
We can provide urlParams to `Discoverlink` by `connect`'ing it to redux.

sqren replied:
This actually works:
```js
export function DiscoverLink({ query, children }: Props) {
  // TODO: Remove hardcoded values for urlParams
  const rangeFrom = 'now-24h';
  const rangeTo = 'now';

  const queryEncoded = {
    _a: rison.encode(query._a),
    _g: rison.encode({
      ...query._g,
      time: {
        from: rangeFrom,
        to: rangeTo
      }
    })
  };

  const pathname = '/app/kibana';
  const search = `_a=${queryEncoded._a}&_g=${queryEncoded._g}`;
  const hash = '/discover';
  const href = getHref(pathname, hash, search);

  return <EuiLink href={href} children={children} />;
}

function getHref(pathname, hash, search) {
  return url.format({
    pathname: chrome.addBasePath(pathname),
    hash: `${hash}?${search}`
  });
}
```

sqren commented on x-pack/plugins/apm/public/store/urlParams.ts:
```diff
@@ -19,6 +20,27 @@ import { IReduxState } from './rootReducer';
+  };
+}
+
+const INITIAL_STATE = calculateTimePickerDefaults();
```
nit: since the input values are constants it's not necessary to check the parsed result, and something a bit shorter might suffice:

```js
const INITIAL_STATE = {
    start: datemath.parse(TIMEPICKER_DEFAULTS.rangeFrom).toISOString(),
    end: datemath.parse(TIMEPICKER_DEFAULTS.rangeTo, { roundUp: true }).toISOString()
}
```

jasonrhodes replied:
`datemath.parse` can return undefined so you have to guard against undefined to satisfy the TS gods.

sqren replied:
Ah.. makes sense.

sqren replied:
(although in this case, it would be better if `datemath.parse` would throw on invalid input, instead of return `undefined`)

jasonrhodes replied:
Agreed, I've had to do this check in a lot of places for what seems like no reason.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3059/)

sqren commented on x-pack/plugins/apm/public/store/urlParams.ts:
```diff
@@ -19,6 +20,27 @@ import { IReduxState } from './rootReducer';
+  };
+}
+
+const INITIAL_STATE = calculateTimePickerDefaults();
```
```suggestion
const INITIAL_STATE: IUrlParams = calculateTimePickerDefaults();
```

jasonrhodes replied:
Good idea, this required a little more type massaging so I took care of it. Thanks.

sqren commented:
Contradicting myself a bit from earlier I think it would be a good idea to debounce the updates when the date picker changes. I noticed quite a bit of like when changing it and don't think it's an edge case anymore.

jasonrhodes committed:
Adds IUrlParams type to initial state with correct typing

jasonrhodes committed:
Updated rison-related snapshots

jasonrhodes committed:
Resolves failing tests related to query param management

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3173/)

jasonrhodes committed:
Adds more tests for Kibana Link and Kibana Rison Link components

jasonrhodes committed:
Re-enables the update button for the EUI super date picker

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3186/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3206/)

jasonrhodes committed:
Adds more Discover link tests

jasonrhodes commented:
> Contradicting myself a bit from earlier I think it would be a good idea to debounce the updates when the date picker changes. I noticed quite a bit of lag when changing it and don't think it's an edge case anymore.

I think now that I've added the update button back (per @formgeist), this problem should be resolved on its own for now.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3374/)

jasonrhodes commented:
@formgeist the update button is re-enabled: https://github.com/elastic/kibana/pull/31311/commits/8a7453f3e892d313337524dc3fbf257a00b88b75

Integration tests for our rison handling are now pushed as well, for the discover links. Would be nice to do something similar for ML links. That work is here: https://github.com/elastic/kibana/pull/31311/commits/c8ef703075cd15c70f086b9493090785fcedd4d5

formgeist commented:
thanks @jasonrhodes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3421/)

jasonrhodes committed:
Moved getRenderedHref to testHelpers, switched Discover Links integration tests to TS, and added ML link integration test

jasonrhodes commented:
> Did you try something like this?
> 
> ```ts
> import * as savedObjects from 'x-pack/plugins/apm/public/services/rest/savedObjects';
> spyOn(savedObjects, getAPMIndexPattern).mockResolvedValue({ id: 'apm-index-pattern-id' });
> ```

@sqren good idea! I always forget jest spies can be stubbed like this. Still sucks that `jest.mock()` is broken in TS test files but I swapped this in and added a comment about that. 

I also added an ML link integration test of the same kind, so I now feel good about our Rison handling and refactoring it whenever we want. :+1:

https://github.com/elastic/kibana/pull/31311/commits/31e32ff15b3bdc413ddb34a770f43272b50e6640

jasonrhodes committed:
Changes how getRenderedHref works to make it clearer where location state is coming from

jasonrhodes committed:
Updates obsolete snapshot

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3617/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3620/)

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/index.tsx:
```diff
@@ -5,13 +5,15 @@
   return {
-    mlAvailable: selectIsMLAvailable(state)
+    mlAvailable: selectIsMLAvailable(state),
+    serviceDetails: getServiceDetails(state)
```
Have you thought about how we can achieve this in the future without Redux?
Eg. if we just rely on hooks inside components to do data fetching, it would be nice if we can avoid global state

sqren replied:
Did you look into lifting the data fetching up so we don't have to access global state?

jasonrhodes replied:
I haven't thought too much about this yet but yeah we'll have to find ways to solve this. I assumed it would be a combination of moving some state around and using Context for things we still want to share throughout the app without having to pass them through 2+ layers etc.

sqren commented on x-pack/plugins/apm/public/components/shared/FilterBar/__test__/DatePicker.test.tsx:
```diff
@@ -0,0 +1,133 @@
+
+    afterAll(() => {
+      global.Date.now = realDateNow;
+    });
```
I think this reads pretty well. I noticed that it was failing on CI with this error so digged a bit into it:

```
plugins/apm/public/components/shared/FilterBar/__test__/DatePicker.test.tsx:37:7 
Type 'Date' is not assignable to type 'number'.

global.Date.now = jest.fn(() => fakeNow);
```

Another approach is to use `spyOn` - that way you can leave the cleanup to jest:
```ts
beforeAll(() => {
  const fakeNow = new Date('2019-02-15T12:00:00.000Z').getTime();
  jest.spyOn(global.Date, 'now').mockReturnValue(fakeNow);
});
```

You could even put it in testHelpers if we need it elsewhere:
```ts
function mockDate(date: string) {
  const fakeNow = new Date(date).getTime();
  jest.spyOn(global.Date, 'now').mockReturnValue(fakeNow);
}
```

jasonrhodes replied:
Yeah this way of mocking time isn't foolproof. I don't know if using jest makes anything better here, though. You still have to use `mockRestore()` or one of the other clearing methods to reset the spy, I think?

jasonrhodes replied:
The bug there was that I was using the Date object for the result of `Date.now` instead of calling `getTime()` to produce the number.

sqren replied:
> The bug there was that I was using the Date object for the result of Date.now instead of calling getTime() to produce the number.

Yes, I just thought it looked nicer if we could avoid doing the cleanup manually. I seem to recall jasmine's implementation did this but perhaps jest went away from that.

sorenlouv reviewed (approved):
LGTM. Nice test coverage 👍 
Let's see if we can remove the last bit of Rison in a follow-up PR.

sqren commented on x-pack/plugins/apm/public/components/shared/Links/url_helpers.test.tsx:
```diff
@@ -28,95 +28,83 @@ describe('fromQuery', () => {
+  //   expect(_a).toEqual(
+  //     `(interval:auto,query:(language:lucene,query:'context.service.name:\"myServiceName\" AND error.grouping_key:\"myGroupId\"'),sort:('@timestamp':desc))`
+  //   );
+  // });
```
Should this be deleted?

jasonrhodes replied:
lol yes

jasonrhodes committed:
Fixes typescript-discovered errors and type errors

jasonrhodes committed:
Finishes up url_helpers tests, removes dead commented tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3879/)

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3911/)

jasonrhodes commented:
Hoping a rebase fixes the func test that failed twice in a row 🤞 

```
12:31:53                  └- ✖ fail: "apis SAML security SAML authentication API access with expired access token. expired access token should be automatically refreshed"
12:31:53                  │      Error: expected 400 "Bad Request", got 200 "OK"
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3929/)

jasonrhodes commented:
Ooh a fun new one:
```
13:24:53 Build step 'Google Cloud Storage Plugin' marked build as failure
13:24:53 ERROR: Step ‘Publish JUnit test result report’ failed: no workspace for elastic+kibana+pull-request/JOB=kibana-ciGroup4,node=immutable #3929
```

jasonrhodes commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3960/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4055/)

jasonrhodes committed:
Removes temporary date picker types from APM

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4059/)

jasonrhodes committed:
Fixes common case for an existing URL bug by not encoding range params

spalger commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4154/)

sqren commented on x-pack/plugins/apm/public/components/shared/Links/url_helpers.ts:
```diff
@@ -16,7 +16,19 @@ export function toQuery(search?: string): APMQueryParamsRaw {
+
+  return qs.stringify(encoded, '&', '=', {
+    encodeURIComponent: (value: string) => value
+  });
```
I'm not even sure if I like the follow proposal but at least it's super explicit (which I do like):

```ts
export function fromQuery2(query: QueryParams) {
  const {
    detailTab,
    flyoutDetailTab,
    kuery,
    page,
    sortDirection,
    sortField,
    spanId,
    traceId,
    transactionId,
    waterfallItemId
  } = query;

  const search = [];

  if (detailTab) {
    search.push(`detailTab=${detailTab}`);
  }

  if (flyoutDetailTab) {
    search.push(`flyoutDetailTab=${flyoutDetailTab}`);
  }

  if (kuery) {
    search.push(`kuery=${encodeURIComponent(kuery)}`);
  }

  // ...

  return search.join('&');
}
```

This way it's very clear how each param is encoded.
